File: claymore-nofee-proxy.py
import socket
import select
import threading
import time
import socketserver
import json
import sys
from collections import OrderedDict

pool_address = 'fakepool'
pool_port = 9999
wallet='0xfakeaddress'
worker_name=''

my_lock = threading.Lock()

def lock_print(msg):
    my_lock.acquire()
    try:
        print("[%s] %s" % (time.ctime(), msg))
    finally:
        my_lock.release()


class ThreadingTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    pass

# modify any requests destined for the remote host
def remove_devfee(raw_data):
    data = None
    try:
        data = bytes.decode(raw_data)
    except:
        lock_print('Unable to decode data: {}'.format(raw_data))
    if data == None:
        return raw_data
    #If it is an Auth packet
    if ('submitLogin' in data) or ('eth_login' in data):
        json_data = json.loads(data, object_pairs_hook=OrderedDict)
        lock_print('[+] Auth in progress with address: ' + json_data['params'][0])
        #If the auth contain an other address than our
        if wallet not in json_data['params'][0]:
             lock_print('[*] DevFee Detected - Replacing Address')
             lock_print('[*] OLD: ' + json_data['params'][0])
             #We replace the address
             json_data['params'][0] = wallet + worker_name
             lock_print('[*] NEW: ' + json_data['params'][0])

        data = json.dumps(json_data) + '\n'

    #Packet is forged, ready to send.
    return str.encode(data)


class StratumProxy(socketserver.StreamRequestHandler):
    def handle_tcp(self, sock, remote):
        try:
            fdset = [sock, remote]
            while True:
                r, w, e = select.select(fdset, [], [])
                if sock in r:
                    r_data = sock.recv(4096)
                    if remote.send(remove_devfee(r_data)) <= 0:
                        break
                if remote in r:
                    if sock.send(remote.recv(4096)) <= 0:
                        break
        finally:
            remote.close()
            lock_print('Close connection: {}'.format(sock.getpeername()))

    def handle(self):
        try:
            sock = self.connection
            lock_print('New Connection: {}'.format(sock.getpeername()))
            remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            remote.connect((pool_address, pool_port))
            self.handle_tcp(sock, remote)
        except socket.error as e:
            lock_print('socket error')
            lock_print(e)


def main():
    # cursory check of command line args
    if len(sys.argv[1:]) != 4:
        lock_print("Usage: ./proxy.py [localport] [remotehost] [remoteport] [ETH Wallet]")
        lock_print("Example: ./proxy.py 9999 eth.realpool.org 9999 0x...")
        sys.exit(0)

    local_port = int(sys.argv[1])

    global pool_address, pool_port
    pool_address = sys.argv[2]
    pool_port = int(sys.argv[3])

    # Set the wallet
    global wallet 
    wallet = sys.argv[4]
    
    lock_print("Wallet set: " + wallet + worker_name)

    lock_print('Starting stratum proxy at port %d' % local_port)
    server = ThreadingTCPServer(('', local_port), StratumProxy)
    server.serve_forever()

if __name__ == '__main__':
    main()

File: claymore-nofee-proxy.go
package main
import(
    "os"
    "fmt"
    "log"
    "net"
    "strings"
    "sync/atomic"
    "encoding/json"
)

var lock_account = ""
var local_port = ""
var remote_address = ""
var remote_port = ""

var conn_num int32 = 0

func main() {
    if len(os.Args) < 5 {
        fmt.Println("Usage: ./proxy [localport] [remotehost] [remoteport] [ETH Wallet]")
        fmt.Println("Example: ./proxy 9999 eth.realpool.org 9999 0x...")
        return
    }
    local_port = os.Args[1]
    remote_address = os.Args[2]
    remote_port = os.Args[3]
    lock_account = os.Args[4]

    log.Println("Wallet set:", lock_account)

    l, err := net.Listen("tcp", ":" + local_port)
    if err != nil {
        log.Fatal("Listen error:", err)
    }
    log.Println("Start proxy at port:", local_port)
    // Loop forever
    for {
        c, err := l.Accept()
        if err != nil {
            log.Println("Accept error:", err)
            continue
        }
        create_proxy(c)
    }
}

func create_proxy(client net.Conn) {
    server, err := net.Dial("tcp", remote_address + ":" + remote_port)
    if err != nil {
        log.Println("Connect to pool error:", err)
        // Close the exist socket
        client.Close()
        return
    }
    atomic.AddInt32(&conn_num, 1)
    log.Println("New connection:", client.RemoteAddr(), " Connection number:", atomic.LoadInt32(&conn_num))
    go handle_conn(client, server, true)
    go handle_conn(server, client, false)
}

func handle_conn(c1, c2 net.Conn, local2server bool) {
    var map_result map[string] interface {}
    buf := make([]byte, 512)
    defer c2.Close()
    defer c1.Close()
    if local2server {
        // Reduce connection number
        defer atomic.AddInt32(&conn_num, -1);
        defer log.Println("Close connection:", c1.RemoteAddr())
    }
    for {
        data_len, err := c1.Read(buf)
        data := buf
        if err != nil {
            // Reduce error log, this case should be client close the socket
            if  err_str := err.Error(); strings.Contains(err_str, "EOF") || strings.Contains(err_str, "use of closed network connection") {
                return
            }
            log.Println("Read error:", err)
            return
        }
        if local2server {
            err = json.Unmarshal(buf[:data_len], &map_result)
            if err != nil {
                // Garbage data, not from claymore
                log.Println("Decode error:", err)
                return
            }
            // Submit eth address
            if v, ok := map_result["method"]; ok && v == "eth_submitLogin" {
                auth_count := map_result["params"].([]interface{})[0].(string)
                log.Println("[*]Auth account:", auth_count)
                if auth_count != lock_account {
                    log.Println("[-]Devfee detected")
                    log.Println("[+]OLD", auth_count)
                    log.Println("[+]NEW", lock_account)
                    buf_str := string(buf[:data_len])
                    data = []byte(strings.Replace(buf_str, auth_count, lock_account, 1))
                    data_len = len(data)
                }
            }
        }
        _, err = c2.Write(data[:data_len])
        if err != nil {
            log.Println("Write error:", err)
            return
        }
    }
}

